src/aiodynamo/fast/credentials.py

```diff
@@ -0,0 +1,306 @@
+    not_expired = auto()
+    expires_soon = auto()
+    expired = auto()
+
+
+EXPIRES_SOON_THRESHOLD = datetime.timedelta(minutes=15)
+EXPIRED_THRESHOLD = datetime.timedelta(minutes=5)
+
+
+@dataclass(frozen=True)
+class Metadata:
+    key: Key
+    expires: datetime.datetime
+
+    def lives_longer_than(self, other: Metadata) -> bool:
+        return self.expires > other.expires
+
+    def check_expiration(self) -> Expiration:
+        """
+
+        :return:
+        """
+        now = datetime.datetime.utcnow()
+        if now >= self.expires:
+            return Expiration.expired
+        diff = self.expires - now
+        if diff < EXPIRED_THRESHOLD:
+            return Expiration.expired
+        if diff < EXPIRES_SOON_THRESHOLD:
+            return Expiration.expires_soon
+        return Expiration.not_expired
+
+
+class MetadataCredentials(Credentials, metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    async def _fetch_metadata(self, http: HTTP) -> Metadata:
+        pass
+
+    async def fetch_with_retry(
+        self,
+        *,
+        http: HTTP,
+        max_attempts: int,
+        url: URL,
+        headers: Optional[Headers] = None
+    ) -> bytes:
+        for attempt in range(max_attempts):
+            try:
+                response = await http.get(url=url, headers=headers)
+            except:
+                logging.exception("GET failed")
+                continue
+            if response:
+                return response
+        raise TooManyRetries()
+
+    def __init__(self):
+        self._metadata: Optional[Metadata] = None
+
+    async def get_key(self, http: HTTP) -> Optional[Key]:
+        await self._check_metadata(http)
+        return self._metadata.key
+
+    @highlander
+    async def _check_metadata(self, http: HTTP):
+        if self._metadata is None:
+            await self._refresh(http)
+            return
+        expiration = self._metadata.check_expiration()
+        if expiration is Expiration.expired:
+            await self._refresh(http)
+            return
+        elif expiration is Expiration.expires_soon:
+            asyncio.create_task(self._maybe_refresh(http))
+            return
+        return self._metadata.key
+
+    @highlander
+    async def _refresh(self, http: HTTP):
+        self._maybe_set_metadata(await self._fetch_metadata(http))
+
+    @highlander
+    async def _maybe_refresh(self, http: HTTP):
+        try:
+            self._maybe_set_metadata(await self._fetch_metadata(http))
+        except:
+            pass
+
+    def _maybe_set_metadata(self, metadata: Metadata):
+        """
+        Guards against `_maybe_refresh` somehow being slower than `_refresh`
+        """
+        if self._metadata is None or metadata.lives_longer_than(self._metadata):
+            self._metadata = metadata
+
+
+class Disabled(Exception):
+    pass
+
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+def and_then(thing: Optional[T], mapper: Callable[[T], U]) -> Optional[U]:
```

Comment:
I can't grok what this does or why it's needed 😓

Reply:
useful in lambdas (below). I'd prefer calling it `map` but that's already taken in Python.